Docs for Unix Domain Sockets (#10252)
hashicorp/consul: website/content/docs/connect/registration/service-registration.mdx

    "destination_service_id": "redis1",
    "local_service_address": "127.0.0.1",
    "local_service_port": 9090,

>     "local_service_socket_path": "/tmp/redis.sock",

Author: @hashicorp/consul-docs team. I've listed this option because 'this shows all the options available'. However it conflicts with local_service_address, so this actual example doesn't work. I think this could generate confusion, but I don't have a great idea how to handle this.

  Defaults to the port advertised by the service instance identified by
  `destination_service_id` if it exists otherwise it may be empty in responses.

> - `local_service_socket_path` - The path of a unix domain socket to connect to the local application

Reviewer: Both `unix` and `Unix` are used — do we have a stylistic guide which recommends one over the other, for consistency?

- `local_bind_address` `(string: "")` - Specifies the address to bind a
  local listener to for the application to make outbound connections to this
  upstream. Defaults to `127.0.0.1`.

> - `local_bind_socket_path` `(string: "")` - Specifies the path to bind a Unix
>   domain socket listener for the application to make outbound connections to
>   this upstream. Defaults to "". This conflicts with specifying the local_bind_port

Reviewer 2: ```suggestion
- `local_bind_socket_path` `(string: "")` - Specifies the path at which to bind a Unix
  domain socket listener for the application to make outbound connections to
  this upstream. This conflicts with specifying the local_bind_port
```

  or local_bind_address. This is only supported when using Envoy as a proxy.

> - `local_bind_socket_mode` `(string: "")` - Specifies an (optional) Unix mode
>   to use for the socket permissions.

Reviewer 2: ```suggestion
- `local_bind_socket_mode` `(string: "")` - Specifies the (optional) Unix octal
  file permissions to use for the socket.
```

>   file permissions to use for the socket. Note this is a string

Reviewer 2: ```suggestion
  file permissions to use for the socket.
```
Is this redundant since the type is mentioned after the option name?

    but the proxy registration will not fail.
  - `protocol` `(string: "http")` - Sets the protocol of the listener. One of `http` or `http2`. For gRPC use `http2`.

### Unix Domain Sockets <sup>Beta</sup>

The following examples show additional configuration for Unix domain sockets.

Added in v1.10.0.

To connect to a service via local Unix Domain Socket instead of a
port, add `local_bind_socket_path` and optionally `local_bind_socket_mode`
to the upstream config for a service:

<Tabs>
<Tab heading="HCL">

```hcl
upstreams = [
  {
    destination_name = "service-1"
    local_bind_socket_path = "/tmp/socket_service_1"
    local_bind_socket_mode = "0700"
  }
]
```

</Tab>
<Tab heading="JSON">

```json
"upstreams": [
  {
      "destination_name":  "service-1",
      "local_bind_socket_path": "/tmp/socket_service_1",
      "local_bind_socket_mode": "0700"
  }
]

</Tab>
</Tabs>

This will cause Envoy to create a socket with the path and mode
provided, and connect that to service-1.

The mode field is optional, and if omitted will use the default mode
for Envoy. This is not applicable for abstract sockets. See

> https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#envoy-v3-api-msg-config-core-v3-pipe

Reviewer 2: ```suggestion
<https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#envoy-v3-api-msg-config-core-v3-pipe>
```

Is this automatically rendered as a link on the site? If not, adding the angle brackets should do this in markdown.

    the listener to be set up. If the port is not free then Envoy will not expose a listener for the path,
    but the proxy registration will not fail.
  - `protocol` `(string: "http")` - Sets the protocol of the listener. One of `http` or `http2`. For gRPC use `http2`.

### Unix Domain Sockets <sup>Beta</sup>

The following examples show additional configuration for Unix domain sockets.

Added in v1.10.0.

To connect to a service via local Unix Domain Socket instead of a
port, add `local_bind_socket_path` and optionally `local_bind_socket_mode`
to the upstream config for a service:

<Tabs>
<Tab heading="HCL">

```hcl
upstreams = [
  {
    destination_name = "service-1"
    local_bind_socket_path = "/tmp/socket_service_1"
    local_bind_socket_mode = "0700"
  }
]
```

</Tab>
<Tab heading="JSON">

> ```json
> "upstreams": [
>   {
>     {
>       "destination_name":  "service-1",
>       "local_bind_socket_path": "/tmp/socket_service_1",
>       "local_bind_socket_mode": "0700"
>     }
>   }
> ]
> ```

Reviewer 2: ```suggestion
```json
"upstreams": [
  {
      "destination_name":  "service-1",
      "local_bind_socket_path": "/tmp/socket_service_1",
      "local_bind_socket_mode": "0700"
  }
]
```
Seems like this was erroneously double nested in a hash.
```

</Tab>
</Tabs>

This will cause Envoy to create a socket with the path and mode
provided, and connect that to service-1.

The mode field is optional, and if omitted will use the default mode
for Envoy. This is not applicable for abstract sockets. See
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#envoy-v3-api-msg-config-core-v3-pipe
for details.

-> These options conflict with the `local_bind_socket_port` and
`local_bind_socket_address` options. For a given upstream the proxy can bind either to an IP port or
a Unix socket, but not both.

Similarly to expose a service listening on a Unix Domain socket to the service
mesh, use either the `socket_path` field in the service definition or the
`local_service_socket_path` field in the proxy definition. These
fields are analogous to the `port` and `service_port` fields in their
respective locations.
<Tabs>
<Tab heading="HCL">

```hcl
services {
  name = "service-2"
  socket_path = "/tmp/socket_service_2"
}
```

</Tab>
<Tab heading="JSON">

```json
"services": {

>   "name": "service-2"
>   "socket_path": "/tmp/socket_service_2"

Reviewer 2: ```suggestion
  "name": "service-2",
  "socket_path": "/tmp/socket_service_2"
```

}
```

</Tab>
</Tabs>
Or in the proxy definition:
<Tabs>
<Tab heading="HCL">

```hcl
services {
  name = "socket_service_2"
  connect {
    sidecar_service {
      proxy {
        name = "service-2"
        local_service_socket_path = "/tmp/socket_service_2"
    ...
      }
    }
  }
}
```

</Tab>
<Tab heading="JSON">

```json
"services": {

>   "name": "socket_service_2"
>   "connect": {

Reviewer 2: ```suggestion
  "name": "socket_service_2",
  "connect": {
```

    the listener to be set up. If the port is not free then Envoy will not expose a listener for the path,
    but the proxy registration will not fail.
  - `protocol` `(string: "http")` - Sets the protocol of the listener. One of `http` or `http2`. For gRPC use `http2`.

### Unix Domain Sockets <sup>Beta</sup>

The following examples show additional configuration for Unix domain sockets.

Added in v1.10.0.

To connect to a service via local Unix Domain Socket instead of a

> port, add local_bind_socket_path and optionally local_bind_socket_mode

Reviewer 2: ```suggestion
port, add `local_bind_socket_path` and optionally `local_bind_socket_mode`
```

to the upstream config for a service:
```json
  {

>     destination_name":  "service-1"
>     local_bind_socket_path = "/tmp/socket_service_1"
>     local_bind_socket_mode = "0700" 

Reviewer 2: ```suggestion
    "destination_name":  "service-1",
    "local_bind_socket_path": "/tmp/socket_service_1",
    "local_bind_socket_mode": "0700" 
```

  }
```
This will cause Envoy to create a socket with the path and mode

> provided, and connect that to service-1

Reviewer 2: ```suggestion
provided, and connect that to service-1.

```

The mode field is optional, and if omitted will use the default mode
for Envoy. This is not applicable for abstract sockets. See
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#envoy-v3-api-msg-config-core-v3-pipe

> for details
> NOTE: These options conflict the local_bind_socket_port and
> local_bind_socket_address options. We can bind to an port or we can
> bind to a socket, but not both.

Reviewer 2: ```suggestion
for details.

-> These options conflict with the `local_bind_socket_port` and
`local_bind_socket_address` options. For a given upstream the proxy can bind either to an IP port or
a Unix socket, but not both.
```

Similarly to expose a service listening on a Unix Domain socket to the service

> mesh use either the 'socket_path' field in the service definition or the
> 'local_service_socket_path' field in the proxy definition. These
> fields are analogous to the 'port' and 'service_port' fields in their

Reviewer 2: ```suggestion
mesh, use either the `socket_path` field in the service definition or the
`local_service_socket_path` field in the proxy definition. These
fields are analogous to the `port` and `service_port` fields in their
```

respective locations.  

> ```json

Reviewer 2: ```suggestion
```hcl
```

  services {
    name = "service-2"
    socket_path = "/tmp/socket_service_2"
  }
```
OR

> ```json

Reviewer 2: ```suggestion
```hcl
```

  proxy {
    name = "service-2"
    local_service_socket_path = "/tmp/socket_service_2"
    ...
  }
```

There is no mode field since the service is expected to create the
socket it is listening on, not the Envoy proxy.

> Again, the socket_path and local_service_socket_path fields conflict
> with address/port and local_service_address/local_service_port
> configuration entries.

Reviewer 2: ```suggestion
Again, the `socket_path` and `local_service_socket_path` fields conflict
with `address`/`port` and `local_service_address`/`local_service_port`
configuration options.
```
